### PR TITLE
Add gige mapper

### PR DESCRIPTION
**Pull request by WangQing-WUT**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it:**
GigE mapper is based on [mapper-sdk-go](https://github.com/kubeedge/mappers-go/tree/main/mapper-sdk-go) and supports GigE Vision protocol cameras access.

**Which issue(s) this PR fixes:**
Fixes https://github.com/kubeedge/mappers-go/issues/71

**Special notes for your reviewer:**
You can find more details in [gige mapper README.md](https://github.com/WangQing-WUT/mappers-go/blob/gige-sdk/mappers/gige/README.md).

### CODE REVIEW

File: mappers/gige-sdk/driver/gigeclient.go [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    void* handle;
+    typedef int (*FPTR)(unsigned int*, char*, char*, char**);
+    handle = dlopen("../bin/librcapi.so", 1);
+    if(handle == NULL){
+        *error = (char *)dlerror();
+        return -1;
+    }
+    FPTR fptr = (FPTR)dlsym(handle, "set_value");
+    if(fptr == NULL){
+        *error = (char *)dlerror();
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigeclient.go:35]: dlclose needed?

File: mappers/gige-sdk/driver/gigeclient.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+		return -1;
+    }
+    int result = (*fptr)(device, feature, value, error);
+	dlclose(handle);
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigeclient.go:39]: alignment. Same issue for all c codes.

File: mappers/gige-sdk/driver/gigeclient.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+	"unsafe"
+)
+
+func (gigEClient *GigEVisionDevice) Set(DeviceSN string, value interface{}) (err error) {
+	convertValue, err := gigEClient.convert(value)
+	if err != nil {
+		return err
+	}
+	if strings.EqualFold(gigEClient.deviceMeta[DeviceSN].FeatureName, "ImageFormat") {
+		if strings.EqualFold(convertValue, "png") {
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigeclient.go:133]: Maybe "switch case" is suitable here.

File: mappers/gige-sdk/driver/gigeclient.go [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+	return results, err
+}
+
+func (gigEClient *GigEVisionDevice) NewClient(DeviceSN string) (err error) {
+	var msg *C.char
+	var dev *C.uint
+	if gigEClient.deviceMeta == nil {
+		gigEClient.deviceMeta = make(map[string]*DeviceMeta)
+	}
+	if _, ok := gigEClient.deviceMeta[DeviceSN]; !ok {
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigeclient.go:240]: if ok {
    return nil
}
if xxx {
    return nil
}

<main body>
return nil

**Review by sailorvii — COMMENTED** [review]
I think packaging the "c" calling to a lib will be better.

File: mappers/gige-sdk/driver/gigeclient.go (hunk 4 above)
Comment (WangQing-WUT) [reply mappers/gige-sdk/driver/gigeclient.go]: Thank you for your valuable comments. The code has been modified. Please review it further.

File: mappers/gige-sdk/rc_genicam_api/set_value.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#include "set_value.h"
+
+extern "C" int set_value(MyDevice myDevice, const char* key, const char* value, char** err) {
+	int ret = SUCCESS;
+	std::string e = "";
+	*err = (char*)malloc(e.length() + 1);
```
Comment (sailorvii) [inline mappers/gige-sdk/rc_genicam_api/set_value.cpp:6]: Put this code to the "try" {}. Remove those "free" in the "catch". Check NULL if malloc failed.

File: mappers/gige-sdk/rc_genicam_api/find_device.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
+#include "find_device.h"
+
+extern "C" void find_device() {
```
Comment (sailorvii) [inline mappers/gige-sdk/rc_genicam_api/find_device.cpp:3]: Not clear what's the meaning. Is it to find devices or print?

File: mappers/gige-sdk/driver/gigedriver.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+			C.close_device(gigEClient.deviceMeta[s].dev)
+			gigEClient.deviceMeta[s].dev = nil
+			gigEClient.deviceMeta[s].deviceStatus = false
+		}
+	}
+	fmt.Println("----------Stop GigE Device Successful----------")
+	return nil
+}
+
+// GetDeviceStatus is an interface to get the device status true is OK , false is DISCONNECTED
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigedriver.go:138]: Add "." after status.

File: mappers/gige-sdk/driver/gigedriver.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	}
+	if visitor != nil {
+		if err = json.Unmarshal(visitor, &gigEClient.visitorConfig); err != nil {
+			fmt.Printf("Unmarshal visitorConfig error: %v\n", err)
+			return "", err
+		}
+	}
+	gigEClient.deviceMeta[gigEClient.protocolCommonConfig.DeviceSN].FeatureName = gigEClient.visitorConfig.FeatureName
+	return gigEClient.protocolCommonConfig.DeviceSN, nil
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigedriver.go:86]: A little weird to return the deviceSN for a function named "set".

File: mappers/gige-sdk/driver/gigedriver.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+// StopDevice is an interface to disconnect a specific device
+func (gigEClient *GigEVisionDevice) StopDevice() (err error) {
+	for s := range gigEClient.deviceMeta {
+		if gigEClient.deviceMeta[s].deviceStatus {
+			C.close_device(gigEClient.deviceMeta[s].dev)
+			gigEClient.deviceMeta[s].dev = nil
+			gigEClient.deviceMeta[s].deviceStatus = false
+		}
+	}
+	fmt.Println("----------Stop GigE Device Successful----------")
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigedriver.go:134]: Replace the fmt.Println with log.

File: mappers/gige-sdk/driver/gigedriver.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+// GetDeviceStatus is an interface to get the device status true is OK , false is DISCONNECTED
+func (gigEClient *GigEVisionDevice) GetDeviceStatus(protocolCommon, visitor, _ []byte) (status bool) {
+	deviceSN, err := gigEClient.SetConfig(protocolCommon, visitor)
+	if err == nil {
+		return false
+	}
+	return gigEClient.deviceMeta[deviceSN].deviceStatus
+}
+
+func (gigEClient *GigEVisionDevice) ReConnectDevice(DeviceSN string) {
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigedriver.go:147]: Reconnect

File: mappers/gige-sdk/driver/gigeclient.go [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+	case uint64:
+		convertValue = strconv.FormatUint(value, 10)
+	case string:
+		convertValue = value
+	case bool:
+		convertValue = strconv.FormatBool(value)
+	case []byte:
+		convertValue = string(value)
+	default:
+		if err == nil {
```
Comment (sailorvii) [inline mappers/gige-sdk/driver/gigeclient.go:362]: Why compare a return value with nil?

File: mappers/gige-sdk/rc_genicam_api/find_device.cpp (hunk 6 above)
Comment (WangQing-WUT) [reply mappers/gige-sdk/rc_genicam_api/find_device.cpp:3]: Find GigE Vision devices and list their device information, such as device name, device serial number, etc.

**Comment by sailorvii** [comment]
Thank you for the change. Rebase all commits into one and no comments from my side. @fisherxu please help to review.

**Review by fisherxu — COMMENTED** [review]
The mapper's name should be `gige`, no need to include the `sdk`, it's not sdk 
cc @WangQing-WUT

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**